### DISCUSSION TITLE
Defining partials of scipy interpolants (interp1d and RegularGridInterpolator)

### DISCUSSION THREAD

**Discussion by TidoHoutepen** [OpenMDAO/OpenMDAO#3203] [Q&A] [answered]
Hi everyone, 

I am building a model of a fuel cell powertrain using an ExplicitComponent to be used by Dymos later in an optimal control problem. I have some outputs that rely on interpolants (specifically interp1d and RegularGridInterpolator) to calculate their efficiencies and the polarisation curve of the fuel cell (i.e. the amount of current drawn for a certain amount of power). 

Let me give an example. The fuel cell in this powertrain is connected to a battery via a boosting DC/DC converter that steps up the voltage of the fuel cell (about 300V) to the voltage of the battery (around say 500V). One of the outputs of my model is the current that charges the battery which, I calculate as follows:

`I = P_FC*eta_DCDC/U_bat`

Where `I` is the output current, `P_FC` is the power the fuel cell produces `U_bat` is the voltage of the battery and finally `eta_DCDC` is the efficiency of the DCDC converter which is calculated using a gridded interpolant that depends on the fuel cell voltage and current (before boosting) calculated using: 

````
# Finally we need to determine the DC/DC operating point
# For this we the current and voltage the FC is operating at
I_FC = FC_current(P_FC)
U_FC = P_FC/I_FC 
op_DCDC = np.vstack([I_FC,U_FC]).T
eta_DCDC = DCDC_eff(op_DCDC)
````
Where `DCDC_eff` is the gridded interpolant and `FC_current` is a 1D interpolant. 

So my question is how do I handle the partials best here? As I can't determine the analytical partials of the interpolants themselves I handled them as constants for now, but that is mathematically incorrect. I also saw something about openMDAO components that are interpolants themselves but I don't know how I would extract their partials either.

**Comment by robfalck** [comment, marked as answer]
I would recommend doing your upstream analysis that computes the inputs for the interpolants, and then using one of the interpolation components that come with OpenMDAO.

For orthogonal data, use [MetaModelStructuredComp](https://openmdao.org/newdocs/versions/latest/features/building_blocks/components/metamodelstructured_comp.html).

This will probably fit your needs, but another possible alternative is the [Surrogate Modeling Toolbox](https://smt.readthedocs.io/en/latest/index.html#focus-on-derivatives), which provides derivatives of the interpolation, including the derivatives of the training data itself, if you require that.

**Reply by TidoHoutepen** [reply]
Thanks for your quick answer! I ended up using the interpND logic as that was the easiest way to integrate it into my component.
